File: dugout-api/src/main/kotlin/com/dugout/api/domain/lineup/dto/LineupDto.kt
package com.dugout.api.domain.lineup.dto

import com.dugout.api.domain.lineup.entity.Lineup
import com.dugout.api.domain.lineup.entity.LineupEntry
import jakarta.validation.Valid
import jakarta.validation.constraints.Max
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import java.time.LocalDateTime

data class LineupEntryPayload(
    @field:NotNull(message = "사용자 ID는 필수입니다")
    val userId: Long,

    @field:NotBlank(message = "포지션은 필수입니다")
    val position: String,

    @field:Min(value = 1, message = "타순은 1 이상입니다")
    @field:Max(value = 9, message = "타순은 9 이하입니다")
    val battingOrder: Int? = null,

    val isBench: Boolean = false,
)

data class SaveLineupRequest(
    @field:Valid
    @field:NotNull(message = "라인업 엔트리는 필수입니다")
    val entries: List<LineupEntryPayload>,
)

data class LineupEntryResponse(
    val id: Long,
    val userId: Long,
    val nickname: String,
    val position: String,
    val battingOrder: Int?,
    val isBench: Boolean,
) {
    companion object {
        fun from(entry: LineupEntry): LineupEntryResponse = LineupEntryResponse(
            id = entry.id,
            userId = entry.user.id,
            nickname = entry.user.nickname,
            position = entry.position,
            battingOrder = entry.battingOrder,
            isBench = entry.isBench,
        )
    }
}

data class LineupResponse(
    val id: Long,
    val matchId: Long,
    val teamId: Long,
    val isAiGenerated: Boolean,
    val isConfirmed: Boolean,
    val confirmedAt: LocalDateTime?,
    val entries: List<LineupEntryResponse>,
) {
    companion object {
        fun of(lineup: Lineup, entries: List<LineupEntry>): LineupResponse = LineupResponse(
            id = lineup.id,
            matchId = lineup.match.id,
            teamId = lineup.team.id,
            isAiGenerated = lineup.isAiGenerated,
            isConfirmed = lineup.isConfirmed,
            confirmedAt = lineup.confirmedAt,
            entries = entries.map(LineupEntryResponse::from),
        )
    }
}

data class LineupRecommendationResponse(
    val matchId: Long,
    val isAiGenerated: Boolean,
    val source: String,            // "STUB" | "AI" — Phase 2 전환 시 "AI"로 변경
    val entries: List<LineupEntryPayload>,
)

File: dugout-ios/Features/Lineup/Sources/Data/DTOs/LineupDTO.swift
//
//  LineupDTO.swift
//  DugoutLineupFeature
//
//  백엔드 LineupResponse / LineupEntryResponse 매핑.
//

import Foundation

struct LineupDTO: Decodable, Sendable {
    let id: Int64
    let matchId: Int64
    let teamId: Int64
    let isAiGenerated: Bool
    let isConfirmed: Bool
    let confirmedAt: Date?
    let entries: [LineupEntryDTO]

    enum CodingKeys: String, CodingKey {
        case id, entries
        case matchId        = "match_id"
        case teamId         = "team_id"
        case isAiGenerated  = "is_ai_generated"
        case isConfirmed    = "is_confirmed"
        case confirmedAt    = "confirmed_at"
    }

    func toDomain() -> Lineup? {
        Lineup(
            id: id,
            matchId: matchId,
            teamId: teamId,
            isAiGenerated: isAiGenerated,
            isConfirmed: isConfirmed,
            confirmedAt: confirmedAt,
            entries: entries.compactMap { $0.toDomain() }
        )
    }
}

struct LineupEntryDTO: Decodable, Sendable {
    let id: Int64
    let userId: Int64
    let nickname: String
    let position: String
    let battingOrder: Int?
    let isBench: Bool

    enum CodingKeys: String, CodingKey {
        case id, nickname, position
        case userId        = "user_id"
        case battingOrder  = "batting_order"
        case isBench       = "is_bench"
    }

    func toDomain() -> LineupEntry? {
        guard let pos = BaseballPosition(rawValue: position) else { return nil }
        return LineupEntry(
            id: id,
            userId: userId,
            nickname: nickname,
            position: pos,
            battingOrder: battingOrder,
            isBench: isBench
        )
    }
}
